python/src/iceberg/transforms.py

```diff
@@ -0,0 +1,520 @@
+        return "null"
+
+    def dedup_name(self) -> str:
+        return self.__str__()
+
+
+class Bucket(Transform):
+    _MURMUR3 = mmh3
+    _MAX_32_BITS_INT = 2147483647
+    _FUNCTIONS_MAP = {  # [0] is hash function and [1] is can_transform check function
+        DateType: (
+            lambda v: Bucket._MURMUR3.hash(struct.pack("q", v)),
+            lambda t: t in [IntegerType, DateType],
+        ),
+        IntegerType: (
+            lambda v: Bucket._MURMUR3.hash(struct.pack("q", v)),
+            lambda t: t in [IntegerType, DateType],
+        ),
+        TimeType: (
+            lambda v: Bucket._MURMUR3.hash(struct.pack("q", v)),
+            lambda t: t in {LongType, TimeType, TimestampType, TimestamptzType},
+        ),
+        TimestampType: (
+            lambda v: Bucket._MURMUR3.hash(struct.pack("q", v)),
+            lambda t: t in {LongType, TimeType, TimestampType, TimestamptzType},
+        ),
+        TimestamptzType: (
+            lambda v: Bucket._MURMUR3.hash(struct.pack("q", v)),
+            lambda t: t in {LongType, TimeType, TimestampType, TimestamptzType},
+        ),
+        LongType: (
+            lambda v: Bucket._MURMUR3.hash(struct.pack("q", v)),
+            lambda t: t in [LongType, TimeType, TimestampType, TimestamptzType],
+        ),
+        StringType: (
+            lambda v: Bucket._MURMUR3.hash(v),
+            lambda t: t == StringType,
+        ),
+        BinaryType: (
+            lambda v: Bucket._MURMUR3.hash(v),
+            lambda t: t == BinaryType or isinstance(t, FixedType),
+        ),
+        UUIDType: (
+            lambda v: Bucket._MURMUR3.hash(
+                struct.pack(
+                    ">QQ",
+                    (v.int >> 64) & 0xFFFFFFFFFFFFFFFF,
+                    v.int & 0xFFFFFFFFFFFFFFFF,
+                )
+            ),
+            lambda t: t == UUIDType,
+        ),
+        # bucketing by Float/Double is not allowed by the spec, but they have hash implementation
+        FloatType: (
+            lambda v: Bucket._MURMUR3.hash(struct.pack("d", v)),
+            lambda t: t == FloatType,
+        ),
+        DoubleType: (
+            lambda v: Bucket._MURMUR3.hash(struct.pack("d", v)),
+            lambda t: t == DoubleType,
+        ),
+    }
+
+    @staticmethod
+    def _decimal_to_bytes(value: Decimal):
+        unscaled_value = Transform._unscale_decimal(value)
+        number_of_bytes = int(math.ceil(unscaled_value.bit_length() / 8))
+        return unscaled_value.to_bytes(length=number_of_bytes, byteorder="big")
+
+    def __init__(self, transform_type: Type, num_buckets: int):
+        if (
+            transform_type not in Bucket._FUNCTIONS_MAP
+            and not isinstance(transform_type, FixedType)
+            and not isinstance(transform_type, DecimalType)
+        ):
+            raise ValueError(f"Cannot bucket by type: {transform_type}")
+
+        super().__init__(
+            f"bucket[{num_buckets}",
```

Comment:
Missing `]`.

Reply:
Good catch. Will update it and also make sure unit test can catch it.